[SINT-408] Test Security of Uptane Partial Verification

cedricvanrompay-datadog opened the pull request:
Making sure Uptane Partial Verification is properly implemented and documented.

cedricvanrompay-datadog committed:
list TODO as comment

cedricvanrompay-datadog committed:
test rejection of unsigned targets

cedricvanrompay-datadog committed:
test rejects invalid signature

cedricvanrompay-datadog committed:
test rejection of revoked keys

also test valid root key rotation

cedricvanrompay-datadog committed:
minor: adding comments

cedricvanrompay-datadog committed:
fix linter error (comment exported values)

cedricvanrompay-datadog committed:
minor: add some documentation

mellon85 commented on pkg/config/remote/uptane/partial_client_test.go:
```diff
@@ -149,9 +151,164 @@ func TestPartialClientVerifyInvalidRotation(t *testing.T) {
+	config.Datadog.Set("remote_configuration.config_root", repository1.configRoot)
+
+	client, err := NewPartialClient()
+	require.NoError(t, err)
```
if you use this more than once I would suggest to do

```
assert := assert.New(t)
```

and 
```
assert.NoError(err)
```

cedricvanrompay-datadog replied:
Done in 85e9be0

mellon85 commented on pkg/config/remote/uptane/partial_client_test.go:
```diff
@@ -149,9 +151,164 @@ func TestPartialClientVerifyInvalidRotation(t *testing.T) {
+	assert.Equal(t, root.Roles["root"].KeyIDs[0], repository2.directorRootKey.PublicData().IDs()[0])
+}
+
+// TestPartialClientRejectsUnsignedTarget tests that the partial uptane client
```
thank you for taking the time to explain the tests!

mellon85 commented on pkg/config/remote/uptane/partial_client_test.go:
```diff
@@ -149,9 +151,164 @@ func TestPartialClientVerifyInvalidRotation(t *testing.T) {
+// does not accept targets which are not listed in the targets metadata file
+func TestPartialClientRejectsUnsignedTarget(t *testing.T) {
+	files := []*pbgo.File{
+		{Path: "datadog/2/APM_SAMPLING/id/1", Raw: []byte("mAlIcIoUs cOnTeNt!!")},
```
can you change `1` to `config`? that's the only configuration file name we have and tests should be as close as possible to reality

cedricvanrompay-datadog replied:
This was copy-pasted from other parts of this file, should I replace `1` with `config` in every other occurrence in this file as well?

mellon85 replied:
I won't ask you to do that, if you want you can

cedricvanrompay-datadog replied:
Done in 0f09576

ameske commented on pkg/config/remote/uptane/partial_client.go:
```diff
@@ -167,7 +170,22 @@ func (c *PartialClient) Update(response *pbgo.ClientGetConfigsResponse) error {
 	err = c.validateAndUpdateTargets(response.Targets.Raw)
 	if err != nil {
-		return err
+		if errors.Is(err, verify.ErrInvalid) {
```
I agree with Dario.

Typed errors are nice when the caller can take different error handling actions based on the class of error. In this case though we're just logging the error in every case so we aren't really handling it any differently.

In addition, if the caller of Update logs the error returned from update (which it likely does) now we have duplicated log information. In a sense, the `PartialClient` should be treated as "library" code and we should let the caller decide how it wants to log error information. We might want to appropriately wrap the error so that there's appropriate context, but I don't think we should log here as well.

cedricvanrompay-datadog replied:
Done in ebb22c5

@ameske : logging? `fmt.Errorf` does not log, it simply wraps the error.

ameske reviewed (commented):
I'd recommend not logging at this part of the code and letting the caller decide how to handle the logging.

The tests look great!

cedricvanrompay-datadog committed:
factorize similar error handling

cedricvanrompay-datadog committed:
(minor) use require.New

cedricvanrompay-datadog committed:
use "config" as filename

resolves https://github.com/DataDog/datadog-agent/pull/11449#discussion_r836569491
